Corrupt serialized resolution result/Problems reading from Binary store when executing `dependencies` tasks or build scans

When executing `./gradlew dependencies --write-locks`, the build fails with this error:

```
integrationStubImplementation - Implementation only dependencies for compilation 'integrationStub' (target  (jvm)).

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':dependencies'.
> Could not resolve all dependencies for configuration ':integrationStubImplementation'.
   > Problems reading data from Binary store in /tmp/gradle15135648821751580614.bin offset 221374 exists? true

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.6/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 3s
1 actionable task: 1 executed

```

When attempting to generate `./gradlew dependencies --scan`, I receive this output:

```

> Task :dependencies FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':dependencies'.
> Could not resolve all dependencies for configuration ':integrationStubImplementation'.
   > Problems reading data from Binary store in /tmp/gradle13858920095948672727.bin offset 261444 exists? true

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.6/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 31s
1 actionable task: 1 executed

A build scan cannot be produced as an error occurred gathering build data.
Please report this problem via https://gradle.com/help/plugin and include the following via copy/paste:

----------
Gradle version: 6.6
Plugin version: 3.4

org.gradle.api.artifacts.ResolveException: Could not resolve all dependencies for configuration ':integrationStubImplementation'.
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.wrapException(ErrorHandlingConfigurationResolver.java:104)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.access$000(ErrorHandlingConfigurationResolver.java:48)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver$ErrorHandlingResolutionResult.getRoot(ErrorHandlingConfigurationResolver.java:203)
        at org.gradle.api.internal.artifacts.configurations.ResolveConfigurationResolutionBuildOperationResult.getRootComponent(ResolveConfigurationResolutionBuildOperationResult.java:59)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.b(SourceFile:116)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.a(SourceFile:70)
        at com.gradle.scan.plugin.internal.b.f.a.h.a(SourceFile:51)
        at com.gradle.scan.plugin.internal.b.f.a.g.a(SourceFile:149)
        at com.gradle.scan.plugin.internal.b.f.c.a(SourceFile:135)
        at com.gradle.scan.plugin.internal.j.a.f.a(SourceFile:12)
        at com.gradle.scan.plugin.internal.j.a$c.finished(SourceFile:154)
        at com.gradle.scan.plugin.internal.j.a.a(SourceFile:65)
        at com.gradle.scan.plugin.internal.j.k.a(SourceFile:58)
        at com.gradle.scan.plugin.internal.j.c.a(SourceFile:103)
        at com.gradle.scan.plugin.internal.j.g.a(SourceFile:48)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:31)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:20)
        at com.gradle.scan.plugin.internal.q.a.c(SourceFile:67)
Caused by: java.lang.RuntimeException: Problems reading data from Binary store in /tmp/gradle13858920095948672727.bin offset 261444 exists? true
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:132)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.lambda$create$0(StreamingResolutionResultBuilder.java:187)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.CachedStoreFactory$SimpleStore.load(CachedStoreFactory.java:101)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:185)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:157)
        at org.gradle.api.internal.artifacts.result.DefaultResolutionResult.getRoot(DefaultResolutionResult.java:48)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver$ErrorHandlingResolutionResult.getRoot(ErrorHandlingConfigurationResolver.java:201)
        at org.gradle.api.internal.artifacts.configurations.ResolveConfigurationResolutionBuildOperationResult.getRootComponent(ResolveConfigurationResolutionBuildOperationResult.java:59)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.b(SourceFile:116)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.a(SourceFile:70)
        at com.gradle.scan.plugin.internal.b.f.a.h.a(SourceFile:51)
        at com.gradle.scan.plugin.internal.b.f.a.g.a(SourceFile:149)
        at com.gradle.scan.plugin.internal.b.f.c.a(SourceFile:135)
        at com.gradle.scan.plugin.internal.j.a.f.a(SourceFile:12)
        at com.gradle.scan.plugin.internal.j.a$c.finished(SourceFile:154)
        at com.gradle.scan.plugin.internal.j.a.a(SourceFile:65)
        at com.gradle.scan.plugin.internal.j.k.a(SourceFile:58)
        at com.gradle.scan.plugin.internal.j.c.a(SourceFile:103)
        at com.gradle.scan.plugin.internal.j.g.a(SourceFile:48)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:31)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:20)
        at com.gradle.scan.plugin.internal.q.a.c(SourceFile:67)
Caused by: java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected module (889) for constraint integrationStubImplementation -> com.vmware.carbonblackcloud:exekute:1.0.0
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.DefaultResolutionResultBuilder.visitOutgoingEdges(DefaultResolutionResultBuilder.java:82)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.deserialize(StreamingResolutionResultBuilder.java:237)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:130)
        ... 21 more
----------

```

Could this be a weird local dependency problem?

Also, I did successfully run ` ./gradlew classes --update-locks com.vmware.carbonblackcloud:exekute`. So I'm not 100% sure if that's just a red herring here or not.

### Expected Behavior

I would like for all of my dependencies to resolve, basically, or flag an actionable error that indicates what is the problem on my end.

### Current Behavior

See the above output, it's pretty much "I can't update my lockfiles using the global command". 

### Context

We're using version locking, and I have several libraries I'm using, and I want to update all dependencies, to look for updates, typically to the AWS library.

### Steps to Reproduce 

This is challenging, because it's in a project with a whole bunch of company IP. And I can't provide a build scan for this task since that part fails, though I'm game for adding any other tasks, debugging, etc., though I bet the debug output would be ok. 

### Your Environment

Linux (Ubuntu 18.04), using JDK 14.0.2, and Gradle 6.6.

This morning, I actually removed that internal project, `exekute` that was listed in the build scan output, and removed some kotlin dependencies from the project. The scan still fails, but with a publicly available POM:

```
BUILD FAILED in 28s
1 actionable task: 1 executed

A build scan cannot be produced as an error occurred gathering build data.
Please report this problem via https://gradle.com/help/plugin and include the following via copy/paste:

----------
Gradle version: 6.6
Plugin version: 3.4

org.gradle.api.artifacts.ResolveException: Could not resolve all dependencies for configuration ':integrationStubImplementation'.
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.wrapException(ErrorHandlingConfigurationResolver.java:104)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.access$000(ErrorHandlingConfigurationResolver.java:48)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver$ErrorHandlingResolutionResult.getRoot(ErrorHandlingConfigurationResolver.java:203)
        at org.gradle.api.internal.artifacts.configurations.ResolveConfigurationResolutionBuildOperationResult.getRootComponent(ResolveConfigurationResolutionBuildOperationResult.java:59)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.b(SourceFile:116)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.a(SourceFile:70)
        at com.gradle.scan.plugin.internal.b.f.a.h.a(SourceFile:51)
        at com.gradle.scan.plugin.internal.b.f.a.g.a(SourceFile:149)
        at com.gradle.scan.plugin.internal.b.f.c.a(SourceFile:135)
        at com.gradle.scan.plugin.internal.j.a.f.a(SourceFile:12)
        at com.gradle.scan.plugin.internal.j.a$c.finished(SourceFile:154)
        at com.gradle.scan.plugin.internal.j.a.a(SourceFile:65)
        at com.gradle.scan.plugin.internal.j.k.a(SourceFile:58)
        at com.gradle.scan.plugin.internal.j.c.a(SourceFile:103)
        at com.gradle.scan.plugin.internal.j.g.a(SourceFile:48)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:31)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:20)
        at com.gradle.scan.plugin.internal.q.a.c(SourceFile:67)
Caused by: java.lang.RuntimeException: Problems reading data from Binary store in /tmp/gradle1615510607535013790.bin offset 220456 exists? true
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:132)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.lambda$create$0(StreamingResolutionResultBuilder.java:187)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.CachedStoreFactory$SimpleStore.load(CachedStoreFactory.java:101)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:185)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:157)
        at org.gradle.api.internal.artifacts.result.DefaultResolutionResult.getRoot(DefaultResolutionResult.java:48)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver$ErrorHandlingResolutionResult.getRoot(ErrorHandlingConfigurationResolver.java:201)
        at org.gradle.api.internal.artifacts.configurations.ResolveConfigurationResolutionBuildOperationResult.getRootComponent(ResolveConfigurationResolutionBuildOperationResult.java:59)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.b(SourceFile:116)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.a(SourceFile:70)
        at com.gradle.scan.plugin.internal.b.f.a.h.a(SourceFile:51)
        at com.gradle.scan.plugin.internal.b.f.a.g.a(SourceFile:149)
        at com.gradle.scan.plugin.internal.b.f.c.a(SourceFile:135)
        at com.gradle.scan.plugin.internal.j.a.f.a(SourceFile:12)
        at com.gradle.scan.plugin.internal.j.a$c.finished(SourceFile:154)
        at com.gradle.scan.plugin.internal.j.a.a(SourceFile:65)
        at com.gradle.scan.plugin.internal.j.k.a(SourceFile:58)
        at com.gradle.scan.plugin.internal.j.c.a(SourceFile:103)
        at com.gradle.scan.plugin.internal.j.g.a(SourceFile:48)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:31)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:20)
        at com.gradle.scan.plugin.internal.q.a.c(SourceFile:67)
Caused by: java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected module (1650) for constraint platform-runtime -> com.rabbitmq:amqp-client:5.9.0
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.DefaultResolutionResultBuilder.visitOutgoingEdges(DefaultResolutionResultBuilder.java:82)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.deserialize(StreamingResolutionResultBuilder.java:237)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:130)
        ... 21 more
----------
```

I'm still zeroing in on what triggers this. I think it may be a platform dependency in a custom source set.

No, unfortunately it still had a problem:

```
Caused by: java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected module (1345) for constraint platform-runtime -> com.rabbitmq:amqp-client:5.9.0
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.DefaultResolutionResultBuilder.visitOutgoingEdges(DefaultResolutionResultBuilder.java:83)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.deserialize(StreamingResolutionResultBuilder.java:238)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:130)
        ... 113 more
```

This was using this version:

```
$ ./gradlew --version

------------------------------------------------------------
Gradle 6.7-20200824220048+0000
------------------------------------------------------------

Build time:   2020-08-24 22:00:48 UTC
Revision:     1c64c1d968445857daf8b3d630f22d20b938fce8

Kotlin:       1.3.72
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.8 compiled on May 10 2020
JVM:          14.0.2 (Oracle Corporation 14.0.2+12-46)
OS:           Linux 5.4.0-42-generic amd64
```

`----------
Gradle version: 6.9.2
Plugin version: 3.7.1
 
org.gradle.api.artifacts.ResolveException: Could not resolve all dependencies for configuration ':ads-engagement-summary-report-job-azkaban_2.11:compileClasspath'.
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.wrapException(ErrorHandlingConfigurationResolver.java:106)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.access$000(ErrorHandlingConfigurationResolver.java:49)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver$ErrorHandlingResolutionResult.getRoot(ErrorHandlingConfigurationResolver.java:215)
        at org.gradle.api.internal.artifacts.configurations.ResolveConfigurationResolutionBuildOperationResult.getRootComponent(ResolveConfigurationResolutionBuildOperationResult.java:59)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.b(SourceFile:113)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.a(SourceFile:66)
        at com.gradle.scan.plugin.internal.b.f.a.h.a(SourceFile:47)
        at com.gradle.scan.plugin.internal.b.f.a.g.a(SourceFile:147)
        at com.gradle.scan.plugin.internal.b.f.c.a(SourceFile:137)
        at com.gradle.scan.plugin.internal.l.a.f.a(SourceFile:12)
        at com.gradle.scan.plugin.internal.l.a$c.finished(SourceFile:154)
        at com.gradle.scan.plugin.internal.l.a.a(SourceFile:65)
        at com.gradle.scan.plugin.internal.l.l.a(SourceFile:58)
        at com.gradle.scan.plugin.internal.l.c.a(SourceFile:98)
        at com.gradle.scan.plugin.internal.l.g.a(SourceFile:51)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:31)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:20)
        at com.gradle.scan.plugin.internal.q.a.c(SourceFile:67)
Caused by: java.lang.RuntimeException: Problems reading data from Binary store in /private/var/folders/ft/b5jkwn010tj3llqw2ljgx1h0001dc1/T/gradle6382969199485670936.bin offset 682978 exists? true
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:132)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.lambda$create$0(StreamingResolutionResultBuilder.java:189)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.CachedStoreFactory$SimpleStore.load(CachedStoreFactory.java:101)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:187)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:159)
        at org.gradle.api.internal.artifacts.result.DefaultResolutionResult.getRoot(DefaultResolutionResult.java:48)
        at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver$ErrorHandlingResolutionResult.getRoot(ErrorHandlingConfigurationResolver.java:213)
        at org.gradle.api.internal.artifacts.configurations.ResolveConfigurationResolutionBuildOperationResult.getRootComponent(ResolveConfigurationResolutionBuildOperationResult.java:59)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.b(SourceFile:113)
        at com.gradle.scan.plugin.internal.b.f.a.h$a.a(SourceFile:66)
        at com.gradle.scan.plugin.internal.b.f.a.h.a(SourceFile:47)
        at com.gradle.scan.plugin.internal.b.f.a.g.a(SourceFile:147)
        at com.gradle.scan.plugin.internal.b.f.c.a(SourceFile:137)
        at com.gradle.scan.plugin.internal.l.a.f.a(SourceFile:12)
        at com.gradle.scan.plugin.internal.l.a$c.finished(SourceFile:154)
        at com.gradle.scan.plugin.internal.l.a.a(SourceFile:65)
        at com.gradle.scan.plugin.internal.l.l.a(SourceFile:58)
        at com.gradle.scan.plugin.internal.l.c.a(SourceFile:98)
        at com.gradle.scan.plugin.internal.l.g.a(SourceFile:51)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:31)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:20)
        at com.gradle.scan.plugin.internal.q.a.c(SourceFile:67)
Caused by: java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected module (2241) for default -> com.linkedin.container-core:container-ic-impl:1.2.14
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.DefaultResolutionResultBuilder.visitOutgoingEdges(DefaultResolutionResultBuilder.java:83)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.deserialize(StreamingResolutionResultBuilder.java:239)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:130)
        ... 21 more
`

Cannot do a successful build anymore after this.

I am getting the same error after upgrading to latest gradle version `8.11.1`. Let me know if there is any other details I can provide:
```
Internal error in Develocity Gradle plugin: finished notification
com.gradle.develocity.DevelocityException: Internal error in Develocity Gradle plugin: finished notification
        at com.gradle.develocity.DevelocityException.a(SourceFile:25)
        at com.gradle.scan.plugin.internal.operations.d.a(SourceFile:138)
        at com.gradle.scan.plugin.internal.operations.h.a(SourceFile:51)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:31)
        at com.gradle.scan.plugin.internal.q.a$a.a(SourceFile:20)
        at com.gradle.scan.plugin.internal.q.a.c(SourceFile:67)
        at java.base@17.0.13/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.RuntimeException: Problems reading data from Binary store in /Users/crogoz/.gradle/.tmp/gradle14438152719818265448.bin offset 5535372 exists? true
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:137)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.lambda$create$0(StreamingResolutionResultBuilder.java:184)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.CachedStoreFactory$SimpleStore.load(CachedStoreFactory.java:101)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:182)
        at org.gradle.api.internal.artifacts.configurations.ResolveConfigurationResolutionBuildOperationResult.getRootComponent(ResolveConfigurationResolutionBuildOperationResult.java:60)
        at java.base/java.util.Optional.map(Optional.java:260)
        at com.gradle.scan.plugin.internal.c.i.a.e$a.b(SourceFile:115)
        at com.gradle.scan.plugin.internal.c.i.a.e$a.a(SourceFile:72)
        at com.gradle.scan.plugin.internal.c.i.a.e.a(SourceFile:51)
        at com.gradle.scan.plugin.internal.c.i.a.d.a(SourceFile:232)
        at com.gradle.scan.plugin.internal.c.i.c.lambda$attach$2(SourceFile:138)
        at com.gradle.scan.plugin.internal.operations.b$c.finished(SourceFile:154)
        at com.gradle.scan.plugin.internal.operations.b.a(SourceFile:65)
        at com.gradle.scan.plugin.internal.operations.n.a(SourceFile:56)
        at com.gradle.scan.plugin.internal.operations.d.a(SourceFile:136)
        ... 5 more
Caused by: java.lang.RuntimeException: Problems loading the resolution results (0.001 secs). Read 2810 values, last was: 5
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.deserialize(StreamingResolutionResultBuilder.java:239)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:135)
        ... 19 more
Caused by: java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected component (346934) for constraint runtimeElements -> <placeholder_groupName>:<placeholder_name>:<version>
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.ResolutionResultGraphBuilder.visitOutgoingEdges(ResolutionResultGraphBuilder.java:144)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.deserialize(StreamingResolutionResultBuilder.java:231)
        ... 20 more


A build scan cannot be produced as an error occurred gathering build data.
Please report this problem via https://gradle.com/help and include the following via copy/paste:

----------
Gradle version: 8.11.1
Plugin version: 3.18
```

This problem appeared to me in Gradle 8.12.1 today due to json-smart versions disappearing from Maven Central (https://github.com/netplex/json-smart-v2/issues/240).

The stack trace was very similar, just mine mentions 'platform' (I have a subproject with `java-platform`).

```
Caused by: java.lang.RuntimeException: Problems reading data from Binary store in /home/pkubowicz/.gradle/.tmp/gradle5551562548940888109.bin offset 0 exists? true
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.DefaultBinaryStore$SimpleBinaryData.read(DefaultBinaryStore.java:137)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.lambda$create$0(StreamingResolutionResultBuilder.java:184)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.store.CachedStoreFactory$SimpleStore.load(CachedStoreFactory.java:101)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.create(StreamingResolutionResultBuilder.java:182)
Caused by: java.lang.IllegalStateException: Corrupt serialized resolution result. Cannot find selected component (6877) for constraint platform-runtime -> net.minidev:json-smart:2.5.1
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.ResolutionResultGraphBuilder.visitOutgoingEdges(ResolutionResultGraphBuilder.java:144)
        at org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.StreamingResolutionResultBuilder$RootFactory.deserialize(StreamingResolutionResultBuilder.java:231)
        ... 135 more
```

Unfortunately, I wasn't able to verify if 8.13 RC fixes the problem, because after Maven Central was fixed, I stopped getting the exception.

--- a/platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ResolutionIssuesIntegrationTest.groovy
+++ b/platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ResolutionIssuesIntegrationTest.groovy
@@ -32,56 +32,6 @@ import spock.lang.Issue
  * file more appropriate for the feature they are testing.
  */
 class ResolutionIssuesIntegrationTest extends AbstractIntegrationSpec {
-
-    @NotYetImplemented
-    @Issue("https://github.com/gradle/gradle/issues/14220#issuecomment-1283947029")
-    def "resolution result represents failure to resolve dynamic selected module version when platform has constraint on that module"() {
-        mavenRepo.module("test", "module1", "11.1.0.1").publish()
-
-        settingsFile << "include 'plat'"
-        file("plat/build.gradle") << """
-            plugins {
-                id("java-platform")
-            }
-
-            dependencies {
-                constraints {
-                    api "test:module1:11.1.0.1"
-                }
-            }
-        """
-
-        buildFile << """
-            plugins {
-                id("jvm-ecosystem")
-            }
-
-            configurations {
-                dependencyScope("implementation")
-                resolvable("runtimeClasspath") {
-                    extendsFrom implementation
-                }
-            }
-
-            tasks.register('resolve') {
-                def root = configurations.runtimeClasspath.incoming.resolutionResult.rootComponent
-                doLast {
-                    println root.get()
-                }
-            }
-
-            ${mavenTestRepository()}
-
-            dependencies {
-                implementation 'test:module1:11.2.0.+'
-                implementation platform(project(":plat"))
-            }
-        """
-
-        expect:
-        succeeds("resolve")
-    }
-
     @NotYetImplemented
     @Issue("https://github.com/gradle/gradle/issues/22326")
     def "capability conflict skip"() {

--- a/platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/constraints/DependencyConstraintsBugsIntegrationTest.groovy
+++ b/platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/constraints/DependencyConstraintsBugsIntegrationTest.groovy
@@ -202,4 +202,65 @@ class DependencyConstraintsBugsIntegrationTest extends AbstractHttpDependencyRes
             }
         }
     }
+
+    def "throwing eachDependency rule on a coordinate carrying only a phantom constraint surfaces the rule's exception"() {
+        // Reproduces this scenario:
+        //   producer: declares a constraint on org:foo:1.1, but does not depend on org:foo
+        //   app: depends on producer and on org:foo:1.0; registers an eachDependency rule that throws
+        //        whenever it sees a non-1.1 version of org:foo
+        //
+        // The throwing rule fires on org:foo:1.0. Historically the rule's exception
+        // was swallowed and the build failed with a confusing internal error referencing the
+        // binary store and an EdgeState with no target component. The rule's own exception should
+        // be surfaced as the resolution failure cause.
+        given:
+        mavenRepo.module("org", "foo", "1.0").publish()
+        mavenRepo.module("org", "foo", "1.1").publish()
+
+        createDirs("producer", "app")
+        settingsFile << """
+            rootProject.name = 'test'
+            include 'producer', 'app'
+        """
+
+        file("producer/build.gradle") << """
+            plugins { id 'java-library' }
+            dependencies {
+                constraints {
+                    api 'org:foo:1.1'
+                }
+            }
+        """
+
+        file("app/build.gradle") << """
+            plugins { id 'java' }
+            repositories {
+                maven { url = "${mavenRepo.uri}" }
+            }
+            dependencies {
+                implementation project(':producer')
+                implementation 'org:foo:1.0'
+            }
+            configurations.all {
+                resolutionStrategy.eachDependency { details ->
+                    if (details.requested.group == 'org' && details.requested.name == 'foo'
+                        && !(details.requested.version ?: '').startsWith('1.1')) {
+                        throw new org.gradle.api.GradleException("enforced version check failed for foo:\${details.requested.version}")
+                    }
+                }
+            }
+            tasks.register('resolveRuntime') {
+                def files = configurations.runtimeClasspath
+                doLast { files.files }
+            }
+        """
+
+        when:
+        fails ':app:resolveRuntime'
+
+        then:
+        failure.assertHasCause("enforced version check failed for foo:1.0")
+        failure.assertHasNoCause("Problems writing to Binary store")
+        failure.assertHasNoCause("No target component for edge")
+    }
 }

--- a/platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/platforms/PlatformResolveIntegrationTest.groovy
+++ b/platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/platforms/PlatformResolveIntegrationTest.groovy
@@ -714,6 +714,54 @@ class PlatformResolveIntegrationTest extends AbstractHttpDependencyResolutionTes
         //Shape of the graph is not checked as bug was failing resolution altogether
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/14220#issuecomment-1283947029")
+    def "resolution result represents failure to resolve dynamic selected module version when platform has constraint on that module"() {
+        mavenRepo.module("test", "module1", "11.1.0.1").publish()
+
+        settingsFile << "\ninclude 'plat'"
+        file("plat/build.gradle") << """
+            plugins {
+                id("java-platform")
+            }
+
+            dependencies {
+                constraints {
+                    api "test:module1:11.1.0.1"
+                }
+            }
+        """
+
+        buildFile << """
+            plugins {
+                id("jvm-ecosystem")
+            }
+
+            configurations {
+                dependencyScope("implementation")
+                resolvable("runtimeClasspath") {
+                    extendsFrom implementation
+                }
+            }
+
+            tasks.register('resolve') {
+                def root = configurations.runtimeClasspath.incoming.resolutionResult.rootComponent
+                doLast {
+                    println root.get()
+                }
+            }
+
+            ${mavenTestRepository()}
+
+            dependencies {
+                implementation 'test:module1:11.2.0.+'
+                implementation platform(project(":plat"))
+            }
+        """
+
+        expect:
+        succeeds("resolve")
+    }
+
     String mavenHttpRepo() {
         """
             repositories {

--- a/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/EdgeState.java
+++ b/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/EdgeState.java
@@ -266,11 +266,21 @@ class EdgeState implements DependencyGraphEdge {
             }
 
             // If we couldn't attach to any nodes, try to inherit any failures that hard edges have
-            // encountered during selection.
+            // encountered during selection. Failures may originate either from target variant
+            // selection (targetNodeSelectionFailure) or from selector resolution itself - for
+            // example when a throwing eachDependency/substitution rule fired on the requested
+            // coordinate and prevented a component from ever being selected.
             if (targetNodes.isEmpty()) {
                 for (EdgeState unattachedEdge : targetComponent.getModule().getUnattachedEdges()) {
-                    if (!unattachedEdge.isConstraint() && unattachedEdge.targetNodeSelectionFailure != null) {
-                        this.targetNodeSelectionFailure = unattachedEdge.targetNodeSelectionFailure;
+                    if (unattachedEdge.isConstraint()) {
+                        continue;
+                    }
+                    ModuleVersionResolveException targetNodeFailure = unattachedEdge.targetNodeSelectionFailure;
+                    if (targetNodeFailure == null && unattachedEdge.selector != null) {
+                        targetNodeFailure = unattachedEdge.selector.getFailure();
+                    }
+                    if (targetNodeFailure != null) {
+                        this.targetNodeSelectionFailure = targetNodeFailure;
                         return;
                     }
                 }

